### DISCUSSION TITLE
classes/class_popupmenu

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#209] [User-contributed notes]
# classes/class_popupmenu

Inherits: Popup< Window< Viewport< Node< Object A modal window used to display a list of options. Description: PopupMenu is a modal window used to display a list of options. Useful for toolbars and...

https://docs.godotengine.org/en/stable/classes/class_popupmenu.html

**Comment by benetherington** [comment]
A handy way to reference PopupMenu items is using an enum. For example:

```gdscript
enum {
    CHECKBOX_ONE,
    CHECKBOX_TWO,
}
var menu: PopupMenu

func _ready() -> void:
    # Create popup menu
    menu = PopupMenu.new()
    menu.add_check_item("Checkbox 1", CHECKBOX_ONE)
    menu.add_separator() # Uh oh! This will cause a problem! See below.
    menu.add_check_item("Checkbox 2", CHECKBOX_TWO)

    # Add signal handler
    menu.id_pressed.connect(_on_popup_item_pressed)

func _on_popup_item_pressed(id: int) -> void:
    # Update item checked state
    var item_idx = menu.get_item_index(id)
    var item_was_checked = menu.get_item_checked(item_idx)
    menu.set_item_checked(idx, not item_was_checked)

    # Run the signal handler for this item
    match id:
        CHECKBOX_ONE:
            _on_checkbox_one_changed(not item_was_checked)
        CHECKBOX_TWO:
            _on_checkbox_two_changed(not item_was_checked)

func _on_checkbox_one_checked(checked: bool) -> void:
    pass

func _on_checkbox_two_checked(checked: bool) -> void:
    pass        

```

If you need to reference PopupMenu items using their ID, make sure you assign an ID to each item, even separators! If you don't, the ID will be set as the length of the menu (`item_count - 1`), and that might match another assigned ID. In the example above, we added three items, but the signal handler function `_on_popup_item_pressed` won't work the way we think it will!

If we run:

```gdscript
for idx in popup.item_count:
	prints(
            "idx", idx,
            "id", popup.get_item_id(idx),
            "text", popup.get_item_text(idx)
        )
```

we will get the output:

```
idx 0 id 0 text Checkbox 1
idx 1 id 1 text
idx 2 id 1 text Checkbox 2
```

You can see that there are two items with an id of 1. When `_on_popup_item_pressed` tries to get the index of CHECKBOX_2 (by using `menu.get_item_index(id)`), it will get the index of the *first* item with that ID, the separator. That means `menu.set_item_checked(idx, not item_was_checked)` won't check or uncheck the second checkbox the way we want it to.

You can solve this problem however you want, but here's a suggestion: since multiple items can have the same ID, why not max out the id value? The intro says `Note: The ID values used for items are limited to 32 bits, not full 64 bits of int. This has a range of -2^32 to 2^32 - 1, i.e. -2147483648 to 2147483647.`, and if you've got a 2147483647-member-long menu enum, you've got bigger problems than id collisions!

Our re-written `_ready` function might look like this:

```gdscript
func _ready() -> void:
    # Create popup menu
    menu = PopupMenu.new()
    menu.add_check_item("Checkbox 1", CHECKBOX_ONE)
    menu.add_separator("", 2147483647) # ahh that's better!
    menu.add_check_item("Checkbox 2", CHECKBOX_TWO)

    # Add signal handler
    menu.id_pressed.connect(_on_popup_item_pressed)
```

You could also set the ID to a lower number you know you won't use, like `99`, but just remember: you can't set it to `-1`, or else the index will be used.

**Reply by ClemensDegasper** [reply]
thanks this helped me alot

**Comment by lfod1997** [comment]
In case anyone needs it: I wrote a [helper method](https://gist.github.com/lfod1997/6e4ad613637d7e8411c1a603999b41d8) (Gist) to help building nested popup menus & updating items' availability.
